m3/m3backend.go

```diff
@@ -0,0 +1,940 @@
+type Protocol int
+
+// Compact and Binary represent the compact and binary thrift protocols respectively
+const (
+	Compact Protocol = iota
+	Binary
+)
+
+const (
+	// MaxTags is the max number of tags per metric
+	MaxTags                 = 10
+	emitMetricBatchOverhead = 19
+	serviceTag              = "service"
+	envTag                  = "env"
+	hostTag                 = "host"
+)
+
+// initializng this in this file's init function to get around lint error
+var maxInt64 int64
+
+func init() {
+	maxInt64 = math.MaxInt64
+}
+
+var (
+	errMaxQueueSize       = errors.New("maxQueueSize must be greater than zero")
+	errMaxPacketSizeBytes = errors.New("maxPacketSizeBytes must be greater than zero")
+	errCommonTagSize      = errors.New("common tag sizes exceeds packet size")
+)
+
+// m3Backend is a metrics backend that reports metrics to the local M3 collector
+// Metrics are buffered, batched up and emitted via the Thrift TCompact protocol over UDP
+type m3Backend struct {
+	client       *m3.M3Client
+	curBatch     *m3.MetricBatch
+	curBatchLock sync.Mutex
+	commonTags   map[*m3.MetricTag]bool
+	freeBytes    int32
+	processors   sync.WaitGroup
+	resourcePool *m3ResourcePool
+	closeChan    chan struct{}
+
+	counters *aggMets
+	gauges   *aggMets
+	timers   *aggMets
+
+	metCh   chan *sizedMetric
+	timerCh chan *timerMetric
+}
+
+// NewM3Backend creates a new M3Backend
+func NewM3Backend(
+	hostPort string,
+	service string,
+	commonTags map[string]string,
+	includeHost bool,
+	maxQueueSize int,
+	maxPacketSizeBytes int32,
+	interval time.Duration,
+) (BufferedBackend, error) {
+	return newMultiM3BackendWithProtocol(backendOptions{
+		hostPorts:          []string{hostPort},
+		service:            service,
+		commonTags:         commonTags,
+		includeHost:        includeHost,
+		maxQueueSize:       maxQueueSize,
+		maxPacketSizeBytes: maxPacketSizeBytes,
+		interval:           interval,
+		protocol:           Compact,
+	})
+}
+
+// NewMultiM3Backend creates a new M3Backend with multiple destinations
+func NewMultiM3Backend(
+	hostPorts []string,
+	service string,
+	commonTags map[string]string,
+	includeHost bool,
+	maxQueueSize int,
+	maxPacketSizeBytes int32,
+	interval time.Duration,
+) (BufferedBackend, error) {
+	return newMultiM3BackendWithProtocol(backendOptions{
+		hostPorts:          hostPorts,
+		service:            service,
+		commonTags:         commonTags,
+		includeHost:        includeHost,
+		maxQueueSize:       maxQueueSize,
+		maxPacketSizeBytes: maxPacketSizeBytes,
+		interval:           interval,
+		protocol:           Compact,
+	})
+}
+
+// NewMultiM3BackendWithProtocol creates a new M3Backend with multiple destinations using the
+// given protocol
+func NewMultiM3BackendWithProtocol(
+	hostPorts []string,
+	service string,
+	commonTags map[string]string,
+	includeHost bool,
+	maxQueueSize int,
+	maxPacketSizeBytes int32,
+	interval time.Duration,
+	protocol Protocol,
+) (BufferedBackend, error) {
```

Comment:
Hey let's turn these three constructors into a single constructor with just an exported option struct to make adding fields not a breaking change. e.g.

```go
backend, err := NewM3Backend(BackendOptions{
  HostPorts: []string{"127.0.0.1:9052"},
  // ..
})
```